quickwit/quickwit-serve/src/metrics.rs

```diff
@@ -17,24 +17,164 @@
 pub struct RestMetrics {
-    pub http_requests_total: IntCounter,
+    /// `http_requests_total` (labels: endpoint, method, status): the total number of HTTP requests
+    /// handled (counter)
+    pub http_requests_total: IntCounterVec<3>,
+    /// `http_requests_pending` (labels: endpoint, method): the number of currently in-flight
+    /// requests (gauge)
+    pub http_requests_pending: IntGaugeVec<2>,
+    /// `http_requests_duration_seconds` (labels: endpoint, method, status): the request duration
+    /// for all HTTP requests handled (histogram)
+    pub http_requests_duration_seconds: HistogramVec<3>,
 }
 
 impl Default for RestMetrics {
     fn default() -> Self {
         RestMetrics {
-            http_requests_total: new_counter(
+            http_requests_total: new_counter_vec(
                 "http_requests_total",
-                "Total number of HTTP requests received",
+                "Total total number of HTTP requests handled (counter)",
                 "quickwit",
+                ["method", "path", "status"],
             ),
+            http_requests_pending: new_gauge_vec(
+                "http_requests_pending",
+                "Number of currently in-flight requests (gauge)",
+                "quickwit",
+                ["method", "path"],
+            ),
+            http_requests_duration_seconds: new_histogram_vec(
+                "http_requests_duration_seconds",
+                "Request duration for all HTTP requests handled (histogram)",
+                "quickwit",
+                ["method", "path", "status"],
+            ),
+        }
+    }
+}
+
+pub type RestMetricsTraceLayer<B> = TraceLayer<
+    SharedClassifier<ServerErrorsAsFailures>,
+    DefaultMakeSpan,
+    RestMetricsRecorder<B>,
+    RestMetricsRecorder<B>,
+    DefaultOnBodyChunk,
+    DefaultOnEos,
+    RestMetricsRecorder<B>,
+>;
+
+/// `RestMetricsRecorder` holds the state(labels) required for recording metrics on a given
+/// request/response.
+pub struct RestMetricsRecorder<B> {
+    pub labels: Arc<Mutex<Vec<String>>>,
+    _phantom: PhantomData<B>,
+}
+
+impl<B> Clone for RestMetricsRecorder<B> {
+    fn clone(&self) -> Self {
+        Self {
+            labels: self.labels.clone(),
+            _phantom: self._phantom,
+        }
+    }
+}
+
+impl<B> RestMetricsRecorder<B> {
+    pub fn new() -> Self {
+        Self {
+            labels: Arc::new(Mutex::new(vec!["".to_string(); 2])),
+            _phantom: PhantomData,
         }
     }
 }
 
-/// Serve counters exposes a bunch a set of metrics about the request received to quickwit.
+impl<B, FailureClass> OnFailure<FailureClass> for RestMetricsRecorder<B> {
+    fn on_failure(
+        &mut self,
+        _failure_classification: FailureClass,
+        _latency: std::time::Duration,
+        _span: &tracing::Span,
+    ) {
+        let labels = self.labels.lock().expect("Failed to unlock labels").clone();
```

Comment:
I was trying to extract `method_and_path` obtaining in a method but it didn't compile. However, the gut feeling says that it should be possible

Reply:
I have not checked, but I suspect the span stores the method and path, and hopefully, we can access those fields later to populate the labels.

Reply:
Thanks for raising this point! 
I've checked by simply putting `dbg!(_span);` in both  `on_failure` and  `on_response` and seems that the structure is barely informative: 
```sh
[quickwit-serve/src/metrics.rs:135] _span = Span {
    name: "request",
    level: Level(
        Debug,
    ),
    target: "tower_http::trace::make_span",
    disabled: true,
    module_path: "tower_http::trace::make_span",
    line: 109,
    file: "/Users/etolbakov/.cargo/index.crates.io-6f17d22bba15001f/tower-http-0.4.4/src/trace/make_span.rs",
}
```